File: Web/NgControllers/courseController.js
app.controller("CourseController", function ($scope, $http) {
    $scope.insertedCourse = {};
    $scope.currentId = null;
    $scope.isEdit = false;

    $scope.getCourses = function () {
        $http.get("/Course/GetCourses")
            .then((res) => {
                $scope.Courses = JSON.parse(res.data);
                console.log($scope.Courses);
            });
    };

    $scope.insertOrEditCourse = function () {
        if ($scope.insertedCourse && $scope.insertedCourse.CourseName && !$scope.isEdit) {
            $http.post("/Course/Insert", $scope.insertedCourse)
                .then(() => {
                    $('#editModal').modal('hide');

                    $scope.clearEditValues();

                    $scope.getCourses();
                });
        } else if ($scope.insertedCourse && $scope.insertedCourse.CourseName && $scope.isEdit) {
            $http.put("/Course/Edit", $scope.insertedCourse)
                .then(() => {
                    $('#editModal').modal('hide');

                    $scope.clearEditValues();

                    $scope.getCourses();
                });
        }
    };

    $scope.deleteCourse = function () {
        if ($scope.currentId) {
            $http.delete("/Course/Delete?id=" + $scope.currentId)
                .then(() => {
                    $('#deleteModal').modal('hide');

                    $scope.currentId = null;

                    $scope.getCourses();
                });
        }
    };

    $scope.getTotalStudents = function (subjects) {
        let totalStudents = 0;

        angular.forEach(subjects, (subject) => {
            angular.forEach(subject.Students, () => {
                totalStudents++;
            })
        });

        return totalStudents;
    };

    $scope.clearEditValues = function () {
        $scope.isEdit = false;
        $scope.insertedCourse = {}
    };

    $scope.setCurrentId = function (currentId) {
        $scope.currentId = currentId;
    };

    $scope.setCourseToEdit = function (course) {
        $scope.insertedCourse.CourseName = course.CourseName;
        $scope.insertedCourse.Id = course.Id;

        $scope.isEdit = true;
    };
})

File: Web/Controllers/CourseController.cs
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Web.Mvc;
using Web.Data;
using Web.Models;

namespace Web.Controllers
{
    public class CourseController : Controller
    {
        private readonly AppDbContext context = new AppDbContext();
        public ActionResult Index()
        {
            return View();
        }

        [HttpGet]
        public ActionResult GetCourses()
        {
            List<Course> courses = context.Courses
                                    .ToList();

            courses.ForEach((course) =>
            {
                decimal? sumGrades = 0.0m;
                decimal count = 0.0m;

                course.CourseSubjects = context.Subjects
                                            .Where(s => s.CourseId == course.Id)
                                            .ToList();

                course.CourseStudents = new List<Student>();
                course.CourseTeachers = new List<Teacher>();

                course.CourseSubjects.ForEach((subject) =>
                {
                    if (course.CourseTeachers.Find(t => t.Id.Equals(subject.TeacherId)) == null) course.CourseTeachers.Add(context.Teachers.FirstOrDefault(s => s.Id == subject.TeacherId));

                    var subjectStudents = context.SubjectStudents
                                            .Where(ss => ss.SubjectId == subject.Id)
                                            .ToList();
                    
                    subjectStudents.ForEach((subjectStudent) =>
                    {
                        if (course.CourseStudents.Find(s => s.Id.Equals(subjectStudent.StudentId)) == null)
                        {
                            course.CourseStudents.Add(context.Students.FirstOrDefault(s => s.Id == subjectStudent.StudentId));
                        }

                        if (subjectStudent.Grade != null)
                        {
                            sumGrades += subjectStudent.Grade;
                            count++;
                        }
                    });
                });

                if (count > 0) course.AverageGrade = sumGrades / count;
            });

            var rtrn = JsonConvert.SerializeObject(courses, Formatting.None, new JsonSerializerSettings()
            {
                ReferenceLoopHandling = ReferenceLoopHandling.Ignore
            });

            return Json(data: rtrn, behavior: JsonRequestBehavior.AllowGet);
        }

        [HttpPost]
        public ActionResult Insert(Course course)
        {
            context.Courses.Add(course);
            context.SaveChanges();

            return new HttpStatusCodeResult(200);
        }

        [HttpPut]
        public ActionResult Edit(Course course)
        {
            var entry = context.Entry(course);
            context.Set<Course>().Attach(course);
            entry.State = EntityState.Modified;

            context.SaveChanges();

            return new HttpStatusCodeResult(200);
        }

        [HttpDelete]
        public ActionResult Delete(int id)
        {
            Course course = context.Courses.FirstOrDefault(c => c.Id == id);

            context.Courses.Remove(course);
            context.SaveChanges();

            return new HttpStatusCodeResult(200);
        }
    }
}